### PR TITLE
Develop crc16

### PR DESCRIPTION
**Pull request by mjdietzx**
Calculate the CRC of the FW image we will send. Then add this to the init packet so it is properly validated. DFU now working end to end for recent revisions of SDK.

Note: I am unsure about importing/exporting modules. May need to double check I did that properly and change some things. Also I think ideally the user shouldn't have to see CRC at all. Rather we would only generate CRC if we are sending the init packet based on the version characteristic as we do in our OPCODE.INITIALIZE_PARAMETERS. I'm not sure how to import the module into dfu.js so maybe you can point me in direction if you think we should do this.

### CODE REVIEW

File: dist/dfu.js [hunk 1]
```diff
@@ -162,10 +162,11 @@
         return view;
     }
 
-    function provision(device, arrayBuffer, imageType) {
+    function provision(device, arrayBuffer, imageType, crc) {
```
Comment (thegecko) [inline dist/dfu.js]: Perhaps consider putting crc ahead of imageType? Is it more likely to send a crc or send a different imageType?

File: example_node.js [hunk 2]
```diff
@@ -40,8 +41,9 @@ dfu.findDevice({ services: [0x180D] })
     var hex = file.toString();
     var buffer = hex2bin(hex);
     log("file length: " + buffer.byteLength);
+    var crc = crc16(buffer);
 
-    return dfu.provision(device, buffer);
+    return dfu.provision(device, buffer, 4, crc);
```
Comment (thegecko) [inline example_node.js]: If we swap the args (above), this negates the need for the '4' here. If we aren't swapping them, you could use `dfu.ImageType.Application` instead

File: example_web.html [hunk 3]
```diff
@@ -68,8 +69,9 @@
                 .then(hex => {
                     var buffer = hex2bin(hex);
                     log("downloaded length: " + buffer.byteLength);
+                    var crc = crc16(buffer);
 
-                    return dfu.provision(device, buffer);
+                    return dfu.provision(device, buffer, 4, crc);
```
Comment (thegecko) [inline example_web.html]: See note above

**Comment by thegecko** [comment]
Added a couple of semantic suggestions, but looks like you nailed the import/export.

**Comment by mjdietzx** [comment]
Agree that is better.

Also what do you think about the user seeing crc16()? Do you think it would be better if we just called this in dfu.js? If we are sending an init packet we know we need the crc so would it be better to call it there? I'm not sure how to do this in javascript while keeping crc as a module if you think we should do this

**Comment by thegecko** [comment]
I could change it to allow an optional crc to be passed in, but without one it generates it for you. That may be the most flexible approach?

**Comment by mjdietzx** [comment]
That sounds like a good approach to me.

**Pull request merged**